File: src/services/audit/detectors/ad/compliance/nist/au2-audit-events.ts
/**
 * NIST AU-2 - Audit Events
 * Checks audit event configuration:
 * - Essential events audited
 * - Audit policy completeness
 */

import { Finding } from '../../../../../../types/finding.types';
import { GpoSecuritySettings } from '../../../../../../providers/smb/smb.provider';

export function detectNistAu2AuditEvents(
  gpoSettings: GpoSecuritySettings | null,
  _includeDetails: boolean
): Finding {
  const issues: string[] = [];

  if (!gpoSettings?.auditPolicies || gpoSettings.auditPolicies.length === 0) {
    issues.push('Audit policy not configured or not readable');
  } else {
    const audit = gpoSettings.auditPolicies;

    // NIST AU-2 required events
    const requiredCategories = [
      'Account Logon',
      'Account Management',
      'Policy Change',
      'System',
      'Object Access',
    ];

    for (const category of requiredCategories) {
      const hasCategory = audit.some(
        (p) => p.category.includes(category) && p.success && p.failure
      );
      if (!hasCategory) {
        issues.push(`${category} not fully audited (success and failure)`);
      }
    }
  }

  return {
    type: 'NIST_AU_2_AUDIT_EVENTS',
    severity: 'medium',
    category: 'compliance',
    title: 'NIST AU-2 - Audit Events Non-Compliant',
    description:
      'Audit event configuration does not comply with NIST AU-2. All security-relevant events should be audited.',
    count: issues.length > 0 ? 1 : 0,
    details: issues.length > 0 ? { violations: issues, framework: 'NIST', control: 'AU-2' } : undefined,
  };
}

File: internal/audit/detectors/ad/compliance/nist/au2-audit-events.go
package nist

import (
	"context"

	"github.com/etcsec-com/etc-collector/internal/audit"
	"github.com/etcsec-com/etc-collector/pkg/types"
)

// AU2AuditEventsDetector checks NIST AU-2 audit events compliance
type AU2AuditEventsDetector struct {
	audit.BaseDetector
}

// NewAU2AuditEventsDetector creates a new detector
func NewAU2AuditEventsDetector() *AU2AuditEventsDetector {
	return &AU2AuditEventsDetector{
		BaseDetector: audit.NewBaseDetector("NIST_AU_2_AUDIT_EVENTS", audit.CategoryCompliance),
	}
}

// Detect executes the detection
func (d *AU2AuditEventsDetector) Detect(ctx context.Context, data *audit.DetectorData) []types.Finding {
	// AU-2 audit event configuration cannot be fully verified via LDAP
	// This requires GPO audit policy analysis

	finding := types.Finding{
		Type:        d.ID(),
		Severity:    types.SeverityMedium,
		Category:    string(d.Category()),
		Title:       "NIST AU-2 Audit Events Review Required",
		Description: "NIST AU-2 audit event configuration cannot be fully verified via LDAP. Manual review of audit policies recommended.",
		Count:       0,
		Details: map[string]interface{}{
			"framework":   "NIST",
			"control":     "AU-2",
			"publication": "SP 800-53",
			"note":        "Audit policy settings require GPO analysis",
			"requiredEvents": []string{
				"AU-2(a): Account logon events",
				"AU-2(a): Account management",
				"AU-2(a): Directory service access",
				"AU-2(a): Logon events",
				"AU-2(a): Object access",
				"AU-2(a): Policy change",
				"AU-2(a): Privilege use",
				"AU-2(a): Process tracking",
				"AU-2(a): System events",
			},
			"recommendations": []string{
				"Enable Advanced Audit Policy Configuration",
				"Audit all security-relevant events",
				"Centralize log collection to SIEM",
				"Set appropriate log retention",
			},
		},
	}

	return []types.Finding{finding}
}

func init() {
	audit.MustRegister(NewAU2AuditEventsDetector())
}
